File: 链表/206.反转链表.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

#
# @lc app=leetcode.cn id=206 lang=python3
#
# [206] 反转链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        return self.reverse(head, None);
    def reverse(self, cur, prev):
        if cur is None:
            return prev
        temp = cur.next
        cur.next = prev
        return self.reverse(temp, cur)
# @lc code=end

File: 链表/206.反转链表.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

/*
 * @lc app=leetcode.cn id=206 lang=cpp
 *
 * [206] 反转链表
 */

// @lc code=start

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        return reverse(head, nullptr);
    }
    
private:
    ListNode* reverse(ListNode* cur, ListNode* prev) {
        if (cur == nullptr) {
            return prev;
        }
        ListNode* temp = cur->next;
        cur->next = prev;
        return reverse(temp, cur);
    }
};
// @lc code=end

